### PR TITLE
Write the script to disk and then run it

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
The protocol now sends the body of the script and not its path. The body needs to written to disk before it can be executed.

EDIT: Because the script needs to written to disk it creates the following to requirements:

* The write should occur asynchronously as the `IO` would otherwise block the daemon, and
* The file needs to be reliable removed without posing a security risk.

Due to the complexity of the above tasks, the `JobRunner` has been upgraded to a `Struct`. This allows for the following:
* The asynchronous `task` and `child` process status now have an object that can be cached on,
* It is easier to validate the `job_id` isn't something nasty (e.g. a relative path to some other directory), and
* The flow of creating the `script` and removing it in a future is clearer.

### CODE REVIEW

File: lib/flight_scheduler/job_runner.rb [hunk 1]
```diff
@@ -39,25 +40,68 @@ module FlightScheduler
...
+      string_envs = envs.map { |k, v| [k.to_s, v] }.to_h
+
+      # Add the job to the registry
+      path = build_script_path
+      FlightScheduler.app.job_registry.add(id, self)
+
+      self.task = Async do
+        # Write the script_body to disk
+        FileUtils.mkdir_p File.dirname(path)
+        File.write path, script_body
```
Comment (WilliamMcCumstie) [inline lib/flight_scheduler/job_runner.rb]: I'm not sure if this should be some form of asynchronous write. If it does then it should be fairly simple to switch out

File: lib/flight_scheduler/job_runner.rb [hunk 2]
```diff
@@ -68,9 +114,8 @@ def run_job(job_id, env, *arguments, **options)
     # * Returns an Async::Task that can be `wait`ed on.  When the returned
     #   task has completed, the subprocess will have been sent a `TERM`
     #   signal.
-    def cancel_job(job_id)
-      process = FlightScheduler.app.job_registry[job_id]
-      if process && process.running?
+    def cancel
+      if child && child.running?
         Async do
           process.kill
```
Comment (benarmston) [inline lib/flight_scheduler/job_runner.rb]: Where does `process` come from?  Should it be `child`?

**Review by benarmston — CHANGES_REQUESTED** [review]
I've added a commit to use `Async::IO::Stream#write` instead of `File.write`.  I'm not sure it makes much difference in practice, but its in keeping with the use of Async.

Job cancellation is also broken I think that's a simple variable name change issue.

File: lib/flight_scheduler/job_runner.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/job_runner.rb]: opps, thought I got them all:
https://github.com/openflighthpc/flight-scheduler-daemon/pull/4/commits/07d6273e71504357f9a4c049532d94783f071744

**Comment by benarmston** [comment]
LGTM

**Pull request merged**